### PR TITLE
Generic breadcrumbs

### PR DESCRIPTION
**Pull request by laymonage**
Fixes #...

_Please check the following:_

-   [ ] Do the tests still pass?[^1]
-   [ ] Does the code comply with the style guide?
    -   [ ] Run `make lint` from the Wagtail root.
-   [ ] For Python changes: Have you added tests to cover the new/fixed behaviour?
-   [ ] For front-end changes: Did you test on all of Wagtail’s supported environments?[^2]
    -   [ ] **Please list the exact browser and operating system versions you tested**:
    -   [ ] **Please list which assistive technologies [^3] you tested**:
-   [ ] For new features: Has the documentation been updated accordingly?

**Please describe additional details for testing this change**.

[^1]: [Development Testing](https://docs.wagtail.org/en/latest/contributing/developing.html#testing)
[^2]: [Browser and device support](https://docs.wagtail.org/en/latest/contributing/developing.html#browser-and-device-support)
[^3]: [Accessibility Target](https://docs.wagtail.org/en/latest/contributing/developing.html#accessibility-targets)

### CODE REVIEW

**Comment by lb-** [comment]
Awesome. Happy to review when ready, I worked a lot on these with Paarth during GSOC last year. 

Is it worth using the wrapper template style for passing 'children' as the 'trailing Breadcrumb title'?

**Comment by laymonage** [comment]
> Awesome. Happy to review when ready, I worked a lot on these with Paarth during GSOC last year.
> 
> Is it worth using the wrapper template style for passing 'children' as the 'trailing Breadcrumb title'?

Thanks, I'm not sure there even is a need to make a special case for the "trailing breadcrumb title", at least for the generic breadcrumbs (maybe not for page breadcrumbs).

In this PR, I simplified the items to be a list of `{"url"?: string; "label": string;}`, where if the `url` is undefined then the item would just be rendered in a `div`, regardless whether it's the last item in the list or not. And the arrow is rendered as long as it's not the final element in the list. So there really isn't a special handling for the "trailing" item: it's the last item in the list, and it can be or not be a link. (It may not make sense for a non-final item to not have a link, but we don't really need to enforce it 🤷)

We can probably simplify the page breadcrumbs implementation to use the same behaviour (maybe even the same template tag, or at least same template) after mapping the `pages` (now `items`) to the `{"url"?: string; "label": string;}` format and do the heavy lifting for the logic in Python.

**Comment by laymonage** [comment]
And in fact there's a bug in the current stable implementation, where if you explore the root page, you'll see `"Root >"` in the breadcrumb, while I believe the `>` shouldn't be there. This is likely due to an edge case in the "trailing breadcrumb title" vs the root-page/for loop counter checking logic.

File: client/tests/stubs.js [hunk 1]
```diff
@@ -62,7 +62,7 @@ class PageChooserModal {}
 global.PageChooserModal = PageChooserModal;
 
 /** Mock window.scrollTo as not provided via JSDom */
-window.scrollTo = jest.fn();
+window.scrollTo = () => {};
```
Comment (laymonage) [inline client/tests/stubs.js:65]: I couldn't use Storybook until I changed this, the console shows "jest is undefined" or something like that because this stubs file is imported in some stories. Seems to be a regression in #10598.

File: wagtail/admin/templates/wagtailadmin/shared/header.html [hunk 2]
```diff
@@ -21,14 +21,12 @@
...
     - `base_actions` - base actions to appear before the main action
     - `extra_actions` - extra action buttons for the header. This is the HTML to be used for the extra buttons
-    - `breadcrumb` - Custom breadcrumbs content as a variable, displayed in place of breadcrumb block before the header content
     - `heading` - Replaces the header `h1` completely with custom variable content
     - `search` - An alternative search content variable displayed if provided and search_url url is not present
     - `extra_rows` - Shown at the end of the header container
 
 {% endcomment %}
 <header class="{% classnames "w-header" classname merged|yesno:"w-header--merged," search_form|yesno:"w-header--hasform," %}">
-    {% block breadcrumb %}{{ breadcrumb }}{% endblock %}
```
Comment (laymonage) [inline wagtail/admin/templates/wagtailadmin/shared/header.html:31]: No longer used

File: wagtail/admin/templates/wagtailadmin/shared/page_breadcrumbs.html
Comment (laymonage) [inline wagtail/admin/templates/wagtailadmin/shared/page_breadcrumbs.html:1]: I believe I didn't change anything in this file other than the comments after extracting the template from the old `breadcrumbs.html`

File: wagtail/contrib/modeladmin/templates/modeladmin/index.html
Comment (laymonage) [inline wagtail/contrib/modeladmin/templates/modeladmin/index.html:1]: Will probably need to update wagtail-modeladmin to match, but I believe it's not necessary as the variable will just be ignored.

File: wagtail/snippets/views/snippets.py [hunk 3]
```diff
@@ -479,6 +507,15 @@ def get_columns(self):
             DateColumn("timestamp", label=_("Date")),
         ]
 
+    def get_breadcrumb_items(self):
+        return self.breadcrumb_items + [
+            {
+                "url": self.get_edit_url(self.object),
+                "label": get_latest_str(self.object),
+            },
+            {"label": _("History")},
```
Comment (laymonage) [inline wagtail/snippets/views/snippets.py:516]: To turn this into a link, the view class needs to define `history_url_name` (its own URL name, basically), and we need to pass the value from the viewset.

I think with our plans of bringing these features to the generic `ModelViewSet`, this should be defined in the generic views. However, to avoid making the scope too big, I'm leaving it for later. This makes the PR focused on creating the generic breadcrumbs tag + using it to refactor the snippets breadcrumbs as an example. Further work will be done in other PRs.

Same case with the usage view.

File: wagtail/snippets/views/snippets.py [hunk 4]
```diff
@@ -1071,6 +1109,18 @@ def get_menu_order(self):
     def get_menu_item_is_registered(self):
         return self.menu_item_is_registered
 
+    @cached_property
+    def breadcrumb_items(self):
```
Comment (laymonage) [inline wagtail/snippets/views/snippets.py:1113]: This allows us to share the first two items for all views instead of duplicating them.

File: wagtail/admin/templates/wagtailadmin/shared/headers/slim_header.html [hunk 5]
```diff
@@ -9,12 +9,18 @@
     <div class="w-pl-slim-header sm:w-pl-5 w-min-h-slim-header sm:w-pr-2 w-w-full w-flex-1 w-overflow-x-auto w-box-border">
         <div class="w-flex w-flex-1 w-items-center w-overflow-hidden">
             {% block header_content %}
+                {% if breadcrumb_items %}
```
Comment (lb-) [inline wagtail/admin/templates/wagtailadmin/shared/headers/slim_header.html:12]: Any chance we can rename this `breadrumbs_items` (plural breadcrumbs)? I see that this will not be a small change now with all the other PRs built on top.

Maybe that can come as a future PR before the release though.

**Review by lb- — APPROVED** [review]
I have gone through each commit locally, thanks for laying this out clearly.

This is approved, **feel free to merge** (I will let you merge in case you want to do any final tidy up of the commits & write the release notes).

There are a few small comments but nothing to block this PR from being merged. Please do reply though so I know whether you think it's worth changing/fixing or have decided otherwise.

## UI Feedback

I am not sure the delete/edit snippet breadcrumbs should provide links to 'itself' as the trailing breadcrumb. We do this in page edit but the link is to the 'view the parent (folder)', and not just a link to the same page you are already on.

Not a blocker for this PR and may already be adjusted in future PRs but I think it makes more sense for these to just be text in the same way the 'new snippet' one is just text and not a link to the `/create'` url.

Screenshot 2023-09-13 at 6 37 32 am
Screenshot 2023-09-13 at 6 38 24 am


## Code feedback

1. Can we reconsider `breadcrum_items` (methods/context etc) and use `breadcrumb_items`. Breadcrumbs being the 'component' (plural), and it having a series of items. This is not critical and it's not worth pushing hard on but just a nice to have if you feel it reads better also.
2. `is_expanded` - I think we can either rename this OR just add a better description in both Python template tag and the template html. In hindsight, this name does not really reflect the concept of 'if this is true, it will be a static set of breadcrumbs that cannot expand or collapse'. Maybe `is_fixed` or `is_static` is better?
3. UI - I am not sure that having 
4. Please confirm that this will not impact ModelAdmin too much, or if it does, you are tracking this on the modeladmin repo. I think it should be all good but just wanted to double check.
5. We really should get our storybook running on the CI, it's not in scope for this PR but I see where we can so easily break the storybook tests without knowing. We have https://github.com/wagtail/wagtail/issues/8660 but not sure if there should be a new issue raised for this task.

File: wagtail/admin/templates/wagtailadmin/shared/header.html (hunk 2 above)
Comment (laymonage) [reply wagtail/admin/templates/wagtailadmin/shared/header.html:31]: The `breadcrumb` block was added in #4029, but then we got rid of ModelAdmin breadcrumbs entirely in #8743, but the block wasn't removed from the `index.html` template (you can see that PR doesn't change `modeladmin/index.html`. Then when #9993 was added, the `breadcrumb` block was retained as an empty fragment.

The same is true for the package, see the following code search:
https://github.com/search?q=repo%3Awagtail-nest%2Fwagtail-modeladmin%20breadcrumb&type=code

**Comment by laymonage** [comment]
> * Can we reconsider `breadcrum_items` (methods/context etc) and use `breadcrumb_items`. Breadcrumbs being the 'component' (plural), and it having a series of items. This is not critical and it's not worth pushing hard on but just a nice to have if you feel it reads better also.

I'm assuming you meant `breadcrumbs_items` as per the other comment. I agree it's better to be consistent, I wanted to use it too but I wasn't sure if the double plural would be OK so I didn't use it. I've amended the commits to use `breadcrumbs_items` and `get_breadcrumbs_items` instead.

> * `is_expanded` - I think we can either rename this OR just add a better description in both Python template tag and the template html. In hindsight, this name does not really reflect the concept of 'if this is true, it will be a static set of breadcrumbs that cannot expand or collapse'. Maybe `is_fixed` or `is_static` is better?

This is technically already in the original template, so I'm inclined to keep it for now. I agree with your point though, and I think we should revisit this when we work on #9498.

> * UI - I am not sure that having

I assume this comment is about the last item in the breadcrumbs? I wasn't really sure about this too. If we want the last item to not be a link to the current page, do we want to enforce this for all (non-page) views e.g. the index view as well? I'm planning to discuss this with Ben (our designer) tomorrow. If you have thoughts on this let me know, in the meantime I'll leave this as-is and will revisit in one of the newer PRs (or open a new one).

> * Please confirm that this will not impact ModelAdmin too much, or if it does, you are tracking this on the modeladmin repo. I think it should be all good but just wanted to double check.

Yep I've tried with my branch of bakerydemo that uses the separate modeladmin package and it works (i.e. no breadcrumbs are shown, and no errors are thrown).

> * We really should get our storybook running on the CI, it's not in scope for this PR but I see where we can so easily break the storybook tests without knowing. We have [Storybook testing: set up Storyshots with automated accessibility tests #8660](https://github.com/wagtail/wagtail/issues/8660) but not sure if there should be a new issue raised for this task.

I agree, having the setup described in that issue would be great to have, but if that takes a lot of work it might be worth having a simpler setup that just runs the Storybook and ensure it doesn't crash.

Thank you for the review, I'll merge this and #10870 for now, hopefully can get to your other points soon.
